ATS.html.markdown

```diff
@@ -0,0 +1,607 @@
+    fun rev_helper {i:nat} (xs: list(a, n-i), acc: list(a, i)): list(a, n) =
+      case xs of
+      | list_nil() => acc
+      | list_cons(x, rest) => rev_helper(rest, list_cons(x, acc))
+  in rev_helper(xs, list_nil) end
+
+// ATS has three context-dependent namespaces
+// the two 'int's mean different things in this example, as do the two 'n's
+fn namespace_example {n:int} (n: int n): int n = n
+//                      ^^^                         sort namespace
+//                    ^          ^^^ ^   ^^^ ^      statics namespace
+// ^^^^^^^^^^^^^^^^^          ^                  ^  value namespace
+
+// a termination metric can go in .< >.
+// it must decrease on each recursive call
+// then ATS will prove it doesn't infinitely recurse
+fun terminating_factorial {n:nat} .<n>. (n: int n): int =
+  if n = 0 then 1 else n * terminating_factorial (n-1)
+
+
+/*************** Part 3: the LF fragment ****************/
+
+// ATS supports proving theorems in LF (http://twelf.org/wiki/LF)
+
+// Relations are represented by inductive types
+
+// Proofs that the nth fibonacci number is f
+dataprop Fib(n:int, m:int) =
+  | FibZero(0, 0)
+  | FibOne(1, 1)
+  | {n, f1, f2: int} FibInd(n, f1 + f2) of (Fib(n-1,f1), Fib(n-2,f2))
+
+// Proved-correct fibonacci implementation
+// [A] B is an existential type: "there exists A such that B"
+// (proof | value)
+fun fib {n:nat} .<n>. (n: int n): [f:int] (Fib(n,f) | int f) =
+  if n = 0 then (FibZero | 0) else
+  if n = 1 then (FibOne | 1) else
+  let
+    val (proof1 | val1) = fib (n-1)
+    val (proof2 | val2) = fib (n-2)
+  // the existential type is inferred
+  in (FibInd(proof1, proof2) | val1 + val2) end
+
+// Faster proved-correct fibonacci implementation
+fn fib_tail {n:nat} (n: int n): [f:int] (Fib(n,f) | int f) =
+  let
+    fun loop {i:int | i < n} {f1, f2: int} .<n - i>.
+          (p1: Fib(i,f1), p2: Fib(i+1,f2)
+          | i: int i, f1: int f1, f2: int f2, n: int n
+          ): [f:int] (Fib(n,f) | int f) =
+      if i = n - 1
+      then (p2 | f2)
+      else loop (p2, FibInd(p2,p1) | i+1, f2, f1+f2, n)
+  in if n = 0 then (FibZero | 0) else loop (FibZero, FibOne | 0, 0, 1, n) end
+
+// Proof-level lists of ints, of type 'sort'
+datasort IntList = ILNil of ()
+                 | ILCons of (int, IntList)
+
+// ILAppend(x,y,z) iff x ++ y = z
+dataprop ILAppend(IntList, IntList, IntList) =
+  | {y:IntList} AppendNil(ILNil, y, y)
+  | {a:int} {x,y,z: IntList}
+    AppendCons(ILCons(a,x), y, ILCons(a,z)) of ILAppend(x,y,z)
+
+// prfuns/prfns are compile-time functions acting on proofs
+
+// metatheorem: append is total
+prfun append_total {x,y: IntList} .<x>. (): [z:IntList] ILAppend(x,y,z)
+  = scase x of // scase lets you inspect static arguments (only in prfuns)
+  | ILNil() => AppendNil
+  | ILCons(a,rest) => AppendCons(append_total())
+
+
+/*************** Part 4: views ****************/
+
+// views are like props, but linear; ie they must be consumed exactly once
+// prop is a subtype of view
```

Comment:
Isn't "view" a subtype of "prop", not the other way around?  Maybe I'm mistaken, but the Intro to ATS book states that view is a linear version of prop.

Reply:
Nevermind, I misunderstood. I'll delete my comments shortly.